forem/forem
# Configurable weighted feed strategy

## What type of PR is this? (check all applicable)


## Description

I wrote extensively about this approach in [Building and Documenting a Nuanced ActiveRecord Approach - DEV Community 👩‍💻👨‍💻](https://dev.to/devteam/building-and-documenting-a-nuanced-activerecord-approach-3j81).  But a quick highlight:

1) Introduce an `AbExperiment` model for field-testing.
2) Introduce a new experiment for feed strategy.  20% of the time a user will get the new strategy.
3) The new strategy is very customizable, in that there are lots of levers we can collectively adjust.
4) The current lever configuration is aiming to favor what the user has expressed as their interests, but also some considerations for the ecosystem.

Both the inline documentation and commit messages may be useful in navigating this experience.

## Related Tickets & Documents

This PR is independent of #15283; which is the first step to introduce a cached value that we could use in the weighted query.

It is working to resolve forem/rfcs#314 (Feed algorithm optimization experiment - suppress reaction).  Reactions are baked into the current implementation.

This also relates to #15180, which were low-stakes refactors that I was performing to better understand the feed.

## QA Instructions, Screenshots, Recordings

This pull request introduces a new field test ([see field_test gem](https://github.com/ankane/field_test/).  I have introduced the `AbExperiment` model to provide a layer around the field test implementation.  One feature of the `AbExperiment` is that you can specify an `ENV` variable to force using one of the variants (note this short circuits the field test wiring but for QA that's alright).

To force the current strategy, you want to use the following: `AB_EXPERIMENT_FEED_STRATEGY=original bin/startup`).

To force the challenger strategy (e.g. `Articles::Feeds::WeightedQueryStrategy`) you need to use the following: `AB_EXPERIMENT_FEED_STRATEGY=weighted_query_strategy bin/startup`.

One particular challenge is that in the challenger strategy, I've implemented a hard-cutoff for articles: those that were published more than 31 days ago.  So if you have an empty feed in your test, consider the published_at date of your articles.

### UI accessibility concerns?

None

## Added/updated tests?


## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post

## [optional] Are there any post deployment tasks we need to perform?

This change introduces some Datadog tracing to gain some insights into performance.  Those are in the feeds controller and the stories controller.

There's a bit of antics going on for those Datadog traces, so please review and help me think of a better approach.

## [optional] What gif best describes this PR or how it makes you feel?

![zoidberg-futurama](https://user-images.githubusercontent.com/2130/140548683-7495a539-4b65-4fcd-841c-986f3bf0707d.gif)
